includes/class-experimental-abtest.php

```diff
@@ -0,0 +1,164 @@
+/**
+ * A class that interacts with Explat A/B tests.
+ *
+ * This class is experimental. It is a fork of the jetpack-abtest package and
+ * updated for use with ExPlat. These changes are planned to be contributed
+ * back to the upstream Jetpack package. If accepted, this class should then
+ * be superseded by the Jetpack class using Composer.
+ *
+ * This class should not be used externally.
+ *
+ * @package WooCommerce\Payments
+ * @link https://packagist.org/packages/automattic/jetpack-abtest
+ */
+
+namespace WCPay;
+
+/**
+ * This class provides an interface to the Explat A/B tests.
+ *
+ * @internal This class is experimental and should not be used externally due to planned breaking changes.
+ */
+final class Experimental_Abtest {
+	/**
+	 * A variable to hold the tests we fetched, and their variations for the current user.
+	 *
+	 * @var array
+	 */
+	private $tests = [];
+
+	/**
+	 * ExPlat Anonymous ID.
+	 *
+	 * @var string
+	 */
+	private $anon_id = null;
+
+	/**
+	 * ExPlat Platform name.
+	 *
+	 * @var string
+	 */
+	private $platform = 'woocommerce';
+
+	/**
+	 * Whether trcking consent is given.
+	 *
+	 * @var bool
+	 */
+	private $consent = false;
+
+	/**
+	 * Constructor.
+	 *
+	 * @param string $anon_id  ExPlat anonymous ID.
+	 * @param string $platform ExPlat platform name.
+	 * @param bool   $consent  Whether tracking consent is given.
+	 */
+	public function __construct( string $anon_id, string $platform, bool $consent ) {
+		$this->anon_id  = $anon_id;
+		$this->platform = $platform;
+		$this->consent  = $consent;
+	}
+
+	/**
+	 * Retrieve the test variation for a provided A/B test.
+	 *
+	 * @param string $test_name Name of the A/B test.
+	 * @return mixed|null A/B test variation, or null on failure.
+	 */
+	public function get_variation( $test_name ) {
+		if ( ! $this->consent ) {
+			return null;
+		}
+		$variation = $this->fetch_variation( $test_name );
+
+		// If there was an error retrieving a variation, conceal the error for the consumer.
+		if ( is_wp_error( $variation ) ) {
+			return null;
+		}
+
+		return $variation;
+	}
+
+	/**
+	 * Fetch and cache the test variation for a provided A/B test from WP.com.
+	 *
+	 * @param string $test_name Name of the A/B test.
+	 * @return mixed|Automattic\Jetpack\Error A/B test variation, or Automattic\Jetpack\Error on failure.
+	 */
+	protected function fetch_variation( $test_name ) {
+		// Make sure test name exists.
+		if ( ! $test_name ) {
+			return new \WP_Error( 'test_name_not_provided', 'A/B test name has not been provided.' );
+		}
+
+		// Make sure test name is a valid one.
+		if ( ! preg_match( '/^[A-Za-z0-9_]+$/', $test_name ) ) {
+			return new \WP_Error( 'invalid_test_name', 'Invalid A/B test name.' );
+		}
+
+		// Return internal-cached test variations.
+		if ( isset( $this->tests[ $test_name ] ) ) {
+			return $this->tests[ $test_name ];
+		}
+
+		// Return external-cached test variations.
+		if ( ! empty( get_transient( 'abtest_variation_' . $test_name ) ) ) {
```

Comment:
Just for my curiousity if you dont mind: I've compared this with jetpack's abtest code and this looks new, is the purpose of this is to have a transient-based cache to improve performance? Or does it have any other purpose besides that?

Reply:
Also, will it have any unintended effects when an experiment is updated? I.e: If a user might have an outdated experiment stored

Reply:
@ilyasfoo Good questions!

> is the purpose of this is to have a transient-based cache to improve performance?

Yeah it's intended to improve performance but is also a feature of ExPlat. Its API response returns a `ttl` value for how long to cache the results for. In my tests recently the value has been 2 hours. I wanted to make sure we cache the variation to help the performance of WP Admin-- the next experiment needs to use the A/B test to decide how to render the WP menu, so this will avoid making an ExPlat API request on every PHP page load.

Ideally stores will have a hosting environment with a [persistent cache](https://developer.wordpress.org/reference/classes/wp_object_cache/#persistent-caching) set up. For those stores, the transient retrieval will be super fast. For other stores, the transient retrieval is a little slower as WordPress will default to the `options` database table for storage.

> will it have any unintended effects when an experiment is updated?

I can't think of any 🎉  With ExPlat, you wouldn't update an experiment as it would interfere with the measurements. Instead you would disable the experiment, then create a new one with a new test name (the name needs to be unique).

Reply:
That's brilliant. Thanks for sharing! 🙇